src/transform/activation.jl

```diff
@@ -0,0 +1,132 @@
+
+# -------------------------------------------------------------
+# Repurposed from https://github.com/tbreloff/OnlineAI.jl
+# -------------------------------------------------------------
+
+abstract Activation <: Transformation
+
+value(activation::Activation, input::AbstractVector) = Float64[value(activation, Σ) for Σ in input]
+deriv(activation::Activation, input::AbstractVector) = Float64[deriv(activation, Σ) for Σ in input]
+
+function value!(buffer::AbstractVector, activation::Activation, input::AbstractVector)
+    for i in 1:length(buffer)
+        buffer[i] = value(activation, input[i])
+    end
+end
+function deriv!(buffer::AbstractVector, activation::Activation, input::AbstractVector)
+    for i in 1:length(buffer)
+        buffer[i] = deriv(activation, input[i])
+    end
+end
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = x
+f'(x) = 1
+"""
+immutable IdentityActivation <: Activation end
+value(activation::IdentityActivation, input::Real) = input
+deriv(activation::IdentityActivation, input::Real) = 1.0
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = 1 / (1 + exp(-x))
+f'(x) = f(x) * (1 - f(x))
+"""
+immutable SigmoidActivation <: Activation end
+value(activation::SigmoidActivation, input::Real) = 1.0 / (1.0 + exp(-input))
+deriv(activation::SigmoidActivation, input::Real) = (s = value(activation, input); s * (1.0 - s))
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = tanh(x)
+f'(x) = 1 - tanh(x)²
+"""
+immutable TanhActivation <: Activation end
+value(activation::TanhActivation, input::Real) = tanh(input)
+deriv(activation::TanhActivation, input::Real) = 1.0 - tanh(input)^2
+
+# ----------------------------------------------------------------------------
+
+doc"""
+f(x) = x / (1 + |x|)
+f'(x) = 1 / (1 + |x|)²
+"""
+immutable SoftsignActivation <: Activation end
+value(activation::SoftsignActivation, input::Real) = input / (1.0 + abs(input))
+deriv(activation::SoftsignActivation, input::Real) = 1.0 / (1.0 + abs(input))^2
+
+# ----------------------------------------------------------------------------
+
+# Rectified Linear Unit
+doc"""
+f(x) = max(0, x)
+f'(x) = { 1,  x > 0
+        { 0,  x ≤ 0
+"""
+immutable ReLUActivation <: Activation end
+value(activation::ReLUActivation, input::Real) = max(0.0, input)
```

Comment:
This function is not typestable and probably very slow for some types of `input`! For example if `typeof(input) <: Float32` then the return type can not be inferred since it can either be `Float64` (the 0.0) or `Float32`

If would suggest

```julia
value{T<:Number}(activation::ReLUActivation, input::T) = max(zero(T), input)